Tests/UnitTests/Plugins/Geant4/Geant4SurfaceProviderTests.cpp

```diff
@@ -0,0 +1,294 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <boost/test/unit_test.hpp>
+
+#include "Acts/Detector/LayerStructureBuilder.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Plugins/Geant4/Geant4SurfaceProvider.hpp"
+#include "Acts/Utilities/BinningType.hpp"
+
+#include <filesystem>
+#include <memory>
+#include <string>
+
+#include "G4Box.hh"
+#include "G4GDMLParser.hh"
+#include "G4LogicalVolume.hh"
+#include "G4NistManager.hh"
+#include "G4PVPlacement.hh"
+#include "G4RotationMatrix.hh"
+#include "G4SystemOfUnits.hh"
+#include "G4ThreeVector.hh"
+#include "G4Transform3D.hh"
+
+class G4VPhysicalVolume;
+
+const int nLayers = 5;
+const int nChips = 5;
+const int nArms = 2;
+
+const double cellDimX = 0.4 * cm;
+const double cellDimY = 0.3 * cm;
+const double cellDimZ = 0.1 * cm;
+
+const double armOffset = 10 * cm;
+
+const std::filesystem::path gdmlPath = "two-arms-telescope.gdml";
+
+BOOST_AUTO_TEST_SUITE(Geant4SurfaceProvider)
+
+std::tuple<G4VPhysicalVolume*, std::vector<std::string>>
+ConstructGeant4World() {
+  // Get nist material manager
+  G4NistManager* nist = G4NistManager::Instance();
+
+  // Option to switch on/off checking of volumes overlaps
+  //
+  G4bool checkOverlaps = true;
+
+  //
+  // World
+  //
+  G4double worldSizeXY = 50 * cm;
+  G4double worldSizeZ = 50 * cm;
+  G4Material* worldMat = nist->FindOrBuildMaterial("G4_Galactic");
+
+  auto solidWorld = new G4Box("World", 0.5 * worldSizeXY, 0.5 * worldSizeXY,
+                              0.5 * worldSizeZ);
+
+  auto logicWorld = new G4LogicalVolume(solidWorld, worldMat, "World");
+
+  auto physWorld = new G4PVPlacement(nullptr, G4ThreeVector(), logicWorld,
+                                     "World", nullptr, false, 0, checkOverlaps);
+
+  G4Material* material = nist->FindOrBuildMaterial("G4_He");
+
+  std::vector<std::string> names;
+  for (int nArm = 0; nArm < nArms; nArm++) {
+    for (int nLayer = 0; nLayer < nLayers; nLayer++) {
+      for (int nChip = 0; nChip < nChips; nChip++) {
+        int sign = (nArm == 0) ? 1 : -1;
+        double posX = sign * (armOffset + nChip * cm);
+        double posY = 0;
+        double posZ = nLayer * cm;
+        G4ThreeVector pos = G4ThreeVector(posX, posY, posZ);
+        G4ThreeVector dims = G4ThreeVector(cellDimX, cellDimY, cellDimZ);
+
+        int cellId = nChips * nLayers * nArm + nChips * nLayer + nChip;
+        std::string name = "cell" + std::to_string(cellId);
+
+        names.push_back(name);
+
+        // Box cell
+        auto solidCell = new G4Box(name, dims[0], dims[1], dims[2]);
+
+        G4LogicalVolume* cellLogical =
+            new G4LogicalVolume(solidCell, material, "cellSensitive");
+
+        new G4PVPlacement(nullptr, pos, cellLogical, name, logicWorld, false, 0,
+                          true);
+      }
+    }
+  }
+
+  // Write GDML file
+  G4GDMLParser parser;
+  parser.SetOutputFileOverwrite(true);
+  parser.Write(gdmlPath.string(), physWorld);
```

Comment:
Does this work if the file already exists? Also this writes the file to the current working directory. Could this instead use a temporary file path?

Reply:
Yes, it does. Line 102 is there specifically to prevent the parser from throwing "File already exists".  I've decided to leave the file path pointing to the working directory because it seems to be the general practice across the Unit Tests (e.g. all the Json tests will generate the ".json" files in the working directory and leave them there).

Reply:
Ok. We might want to fix this at some point in the future.